### PR TITLE
#3710 Add rule to team

### LINKED ISSUE #3710: [RulesDashboard] - Add Rule to a Team
### Description

We need to be able to assign our rule to any specific teams. This is a post endpoint which should take in a ruleID and array of teamIDs to assign to, and update the rule to store the assigned teams. The endpoint should return a success message or errors.

Be sure to include unit tests and screenshots of postman tests

### Acceptance Criteria

- [ ] new endpoint created under route `rules/rule/:ruleId/assign-team
- [ ] endpoint returns a message
- [ ] unit tests
- [ ] postman

### Proposed Solution

Should not be too complicated, just a simple update to the rule in the database. Make sure to refer to the mocks to understand how we will be using the endpoint before implementation.

### Mocks

Image

### PR DESCRIPTION
**Pull request by chrisdicono**
## Changes

Created the endpoint to add rules to teams.

## Test Cases

- Fails if user is a guest
- Fails if rule does not exist
- Fails if rule is deleted
- Fails if an inputted teamId does not exist
- Successfully assigns a team to a rule's "teams" attribute.

## Screenshots

Screenshot 2025-11-19 174310
Screenshot 2025-11-19 174331
Screenshot 2025-11-19 174500
Screenshot 2025-11-19 174511
Screenshot 2025-11-19 174555
Screenshot 2025-11-19 174603
Screenshot 2025-11-19 174741
Screenshot 2025-11-19 174749
Screenshot 2025-11-19 175117
Below, a rule is deleted to test edge case with deleted rules:
Screenshot 2025-11-19 175423
Screenshot 2025-11-19 175529
Screenshot 2025-11-19 175536

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [x] All commits are tagged with the ticket number
- [x] No linting errors / newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [x] No merge conflicts
- [x] All checks passing
- [x] Screenshots of UI changes (see Screenshots section)
- [x] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [x] No `yarn.lock` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack
- [x] PR is linked to the ticket (fill in the closes line below)

Closes #3710

### CODE REVIEW

File: src/backend/src/controllers/rules.controllers.ts [hunk 1]
```diff
@@ -106,6 +116,19 @@ export default class RulesController {
...
   }
 
+  static async assignRuleTeam(req: Request, res: Response, next: NextFunction) {
+    try {
+      const { ruleId } = req.params;
+      const { teamIds } = req.body;
+
+      await RulesService.assignRuleTeam(ruleId, teamIds, req.currentUser, req.organization);
+
+      res.status(200).json({ message: `Teams of rule ${ruleId} successfully updated.` });
```
Comment (cielbellerose) [inline src/backend/src/controllers/rules.controllers.ts]: we should be returning the whole updated Rule object here

File: src/backend/src/services/rules.services.ts [hunk 2]
```diff
@@ -462,6 +463,57 @@ export default class RulesService {
...
+      include: { teams: true }
+    });
+    if (!rule) {
+      throw new NotFoundException('Rule', ruleId);
+    }
+    if (rule.deletedByUserId) {
+      throw new DeletedException('Rule', ruleId);
+    }
+
+    // Checks that each team exists
```
Comment (cielbellerose) [inline src/backend/src/services/rules.services.ts]: could we check here that the team is in the correct organization - also Team has a dateArchived field so could we ensure the team was not archived

**Review by cielbellerose — CHANGES_REQUESTED** [review]
some notes, but looks great so far! could you remove the delete project rule parts from this ticket? Also completely our bad, but we talked recently and decided the team frontend should only allow users to add rules to one team at a time. Could you update this so it only takes in one teamId?

File: src/backend/src/services/rules.services.ts [hunk 3]
```diff
@@ -462,6 +463,66 @@ export default class RulesService {
...
+      });
+    }
+
+    // retrieve and return the updated rule
+    const newRule = await prisma.rule.findUnique({
+      where: { ruleId },
+      include: { teams: true }
+    });
+
+    return newRule;
```
Comment (cielbellerose) [inline src/backend/src/services/rules.services.ts]: make sure we're passing this through the transformer

File: src/backend/src/services/rules.services.ts [hunk 4]
```diff
@@ -462,6 +463,66 @@ export default class RulesService {
     return projectRuleTransformer(updatedProjectRule);
   }
 
+  /**
+   * Assigns each team in the inputted array of rules to
+   * the rule with the inputted ruleId.
+   * @param ruleId The ruleId of the rule to be added to
+   * @param teamIds An array of teamId's to be added to the rule
```
Comment (cielbellerose) [inline src/backend/src/services/rules.services.ts]: update for one team ID

**Review by cielbellerose — APPROVED** [review]
lgtm! very small updates

File: src/backend/src/services/rules.services.ts [hunk 5]
```diff
@@ -462,6 +463,66 @@ export default class RulesService {
...
+      throw new DeletedException('Rule', ruleId);
+    }
+
+    // Checks based on the team
+    const team = await prisma.team.findUnique({ where: { teamId } });
+    if (!team) throw new NotFoundException('Team', teamId);
+    if (team.organizationId !== org.organizationId) throw new InvalidOrganizationException('Rule');
+    if (team.dateArchived) throw new HttpException(400, 'Cannot assign an archived team.');
+
+    // We add the team to the rule if it is not already in the rule
```
Comment (chpy04) [inline src/backend/src/services/rules.services.ts:509]: if its already on the team, either we should error, or we should make this endpoint toggle it. @cielbellerose what do you think

File: src/backend/src/services/rules.services.ts [hunk 6]
```diff
@@ -462,6 +463,66 @@ export default class RulesService {
...
+    const rule = await prisma.rule.findUnique({
+      where: { ruleId },
+      include: { teams: true }
+    });
+    if (!rule) {
+      throw new NotFoundException('Rule', ruleId);
+    }
+    if (rule.deletedByUserId) {
+      throw new DeletedException('Rule', ruleId);
+    }
```
Comment (chpy04) [inline src/backend/src/services/rules.services.ts:498]: need to ensure the rule is part of the same organization as the user

**Review by chpy04 — CHANGES_REQUESTED** [review]

File: src/backend/src/services/rules.services.ts (hunk 5 above)
Comment (cielbellerose) [reply src/backend/src/services/rules.services.ts:509]: probably toggle. yeah since we call this endpoint on the team page when a rule is clicked, calling this endpoint on a rule thats already in a team would remove the rule from the team - makes sense to me.

File: src/backend/src/routes/rules.routes.ts [hunk 7]
```diff
@@ -44,6 +44,13 @@ rulesRouter.post(
   RulesController.editProjectRuleStatus
 );
 
+rulesRouter.post(
+  '/rule/:ruleId/assign-team',
```
Comment (chpy04) [inline src/backend/src/routes/rules.routes.ts]: if toggling we need to rename the url and all the function names to reflect this

**Review by chpy04 — CHANGES_REQUESTED** [review]

**Review by chpy04 — APPROVED** [review]
